### DISCUSSION TITLE
dbt project with partitioned and unpartitioned assets

### DISCUSSION THREAD

**Discussion by the4thamigo-uk** [dagster-io/dagster#19218] [Q&A] [answered]
Hi I have a job with partitioned (time-series data) and unpartitioned assets (i.e. lookup tables), and I have a dbt job that builds models from these assets. If I specify a `partitions_def` to the `dbt_assets` decorator, then my unpartitioned dbt assets end up being partitioned. Is there a workaround to ensure that my partitioned dagster assets map to partitioned dbt assets, and unpartitioned dagster assets to unpartitioned dbt assets?

**Comment by the4thamigo-uk** [comment, marked as answer]
I worked around this by having two `@dbt_asset` definitions, one with a `partitions_def` and one without. I then used dbt tags to mark which dbt models are partitioned daily (`partitioned_daily`). Then I could assign those models to the partitioned `@dbt_asset` using the `select="tag:partitioned_daily"` parameter. The unpartitioned `@dbt_asset` excludes these models with `exclude="tag:partitioned_daily"`

**Reply by toddy86** [reply]
I ended up going down a similar route with the dbt tags. But for anyone else who stumbles across this, if you then want to use both your partitioned and non-partitioned assets in a e.g. a dbt job section, Dagster will throw you a `Exactly one dbt AssetsDefinition is required` error.

In my use case, I needed to be able to easily pass around and select from the union of both the partitioned and non-partitioned assets. Might be other ways around this, but I simply looped over the `dbt_asset` definitions. Code snippet below in case it helps someone else

```
def dbt_asset_selection(
    dbt_assets: list[list[AssetsDefinition]],
    dbt_select: str,
    dbt_exclude: Optional[str] = None,
) -> Optional[AssetSelection]:
    """
    Utility to allow for selecting dbt assets from multiple dbt asset definitions.
    Dagster presently only allows for the use of a single dbt asset definition in a single asset selection. Therefore,
    we need to loop over the results and return the union of the selection.
    """
    selection = None
    for asset_definition in dbt_assets:
        assets = build_dbt_asset_selection(
            dbt_assets=[*asset_definition],
            dbt_select=dbt_select,
            dbt_exclude=dbt_exclude,
        )
        if selection is None:
            selection = assets
            continue

        selection = selection | assets

    return selection
```

**Reply by toddy86** [reply]
Hmmmm, this approach doesn't seem to work actually. When materializing the assets (either via a job or the asset page), the incremental and full-refresh jobs are spawned as separate steps in the DAG. Resulting in the dbt lineage being broken and asset checks/tests failing (as the other required assets for relationship tests have not yet materialized). 

This _appears_ to be isolated to integrity tests e.g. in step 1, we materialize asset A and run a check that all ID in A appear in asset B. But, asset B isn’t materialized until the next step - so we obviously fail. 

Is there a way to combine dbt partitioned and non-partitioned assets in a job AND have both assets execute in a single step?

Other than removing the integrity test (not ideal), are there any ways around this? Ideal goal is to have all of the dbt assets run in a single step.

@garethbrickman , @rexledesma  or anyone else from the Dagster team. What is the best way to work around this at this time? 

The only idea that I've had so far is to tag each of the integrity tests and exclude them from the run, then at the end of the run, run these tests. Quite a lot of faff.

![image](https://github.com/dagster-io/dagster/assets/10559757/418f1ba9-0258-42e1-af9a-be446aec06d1)

For reference, we are currently constructing our dbt assets like the below, and them using these in the jobs with the above custom code.

```
# build_dbt_assets is a thin wrapper around the @dbt_assets decorator
dbt_assets_partitioned_by_day = build_dbt_assets(
    name="core_partitioned_by_day",
    select=DAILY_PARTITION,
    partitions_def=CORE_DAILY_PARTITION,
    backfill_policy=BackfillPolicy.single_run(),
)
dbt_assets_full_refresh = build_dbt_assets(
    name="core_full_refresh",
    exclude=DAILY_PARTITION,
)

dbt_assets = [dbt_assets_full_refresh, dbt_assets_partitioned_by_day]

# jobs/__init__.py
core_data_job = define_asset_job(
        name="core_data_job",
        selection=dbt_asset_selection(
            dbt_assets=dbt_assets,
            dbt_select="tag:my_tag_1 tag:my_tag_2 tag:my_tag_2",
        ),
        partitions_def=CORE_DAILY_PARTITION,
    )

# We also then pass around the dbt_assets to other assets like the following to allow for dependency resolution
# This uses a similar approach to the dbt_asset_selection code where we have implemented a custom dbt function 
# called dbt_asset key where we loop over the 2x dbt_assets to find the key
@asset(
    key_prefix="my_prefix",
    partitions_def=CORE_DAILY_PARTITION,
    deps=dbt_asset_key(dbt_assets=dbt_assets, model_name="dbt_model_name_to_use"),
)
def asset_1(context: AssetExecutionContext) -> None:
   ...
```

P.S. apologies for the tag, but this is marked as a closed issue and I presume it wouldn't have visibility otherwise - but equally thought it would be best to keep this on a single discussion for everyone else who stumbles across it.

**Reply by OwenKephart** [reply]
@toddy86 Can you say more about how your integrity tests are structured? It sounds like, in essence, you have a test for A which is actually dependent on both A and B.

So, if your dbt_assets function is essentially: run all selected assets, then test all selected assets, this ends up not being resilient to being arbitrarily subset, as there is a sort of "hidden" dependency relationship wherein the test for A depends on B being materialized, and that is not represented anywhere in dagster. Is it all sensible in your case to rethink that test as a test of B (which happens to reference A)? This would allow the test to be executed in the same step as B, resolving this issue.

That's not to say that this is a catch-all solution, I think if the test referenced both A and B, there could still be confusing behavior, depending on the body of your dbt_assets function -- do you mind sharing your execution function?

**Reply by toddy86** [reply]
Hey @OwenKephart , 

Thanks for taking the time to reply. 

> in essence, you have a test for A which is actually dependent on both A and B

This is correct. We are using the dbt [relationships](https://docs.getdbt.com/reference/resource-properties/data-tests#relationships) test to ensure we have relational integrity across a variety of models. An example is below.

```
- name: model_a
    columns:
      - name: table_b_sk
        tests:
          - relationships:
              to: ref('table_b')
              field: sk
```

> Is it all sensible in your case to rethink that test as a test of B (which happens to reference A)? This would allow the test to be executed in the same step as B, resolving this issue.

We can't do this, as unless I am mistaken, the way in which the relationships test works is that is checking for that all of the values in table A exist in table B. Whereas, we can't always have the inverse where all of the values from table B are in table A. 

> do you mind sharing your execution function
Sorry, what exactly do you mean by "execution function"?

My current line of thinking as the simplest (but not the greatest) implementation would be to simply have a sensor which kicks off a job once our core asset job has finished, with this sensor kicking off a job which simply runs all of the relational integrity tests.

At present though, this wouldn't work for us, as we select the assets for the core data job using group names (which are derived from dbt tags + other non-dbt assets which have the same group name). We could probably re-work this a little though if you think this would be the best path forward. 

```
core_data_job = define_asset_job(
        name="core_data_job",
        selection=
            AssetSelection.groups("group_1_with_dbt_and_non_dbt_assets")
            | AssetSelection.groups("group_2_with_dbt_and_non_dbt_assets")
            | AssetSelection.groups("group_3_with_dbt_and_non_dbt_assets")
        partitions_def=CORE_DAILY_PARTITION,
    )
```

**Reply by toddy86** [reply]
Ah, I see this is actually referenced on the [Asset Check discussion](https://github.com/dagster-io/dagster/discussions/16266#discussioncomment-7516205) as something that is a desired feature - which by the looks of it came out early February. But I am guessing this doesn't take into account the above use case where we have partitioned and non-partitioned assets in the same job (and therefore separate step functions)

**Reply by SeppBerkner** [reply]
@toddy86 I am now at the similar step in my workflow, where I want to have relationship tests of my intermediate models before pushing to the data mart. Because my models are created in different sets of jobs (for different reasons) I also came to the conclusion, that I run the `dbt build` on all models excluding the relationship tests with the dbt selection `dbt test -s "test_name:*relationships*"` and have a seperate job + sensor that kicks in when all models have been built and tested. I just wanted to know if you actually implemented it like that for your use case as well and if you think it is a reasonable way or if there has been new ways of dealing with this situation.

**Comment by the4thamigo-uk** [comment]
It looks like there is some recent work to deal with this issue properly and hopefully avoid the need to split up the dbt_asset into separate ones. https://github.com/dagster-io/dagster/pull/26625. However, it is not clear to me whether this initiative has resulted in a production-ready feature yet or not. I wonder if anyone has been following this work?
